### DISCUSSION TITLE
Should MigrationStrategy.onCreate run on connection open? Or first create?

### DISCUSSION THREAD

**Discussion by fgblomqvist** [simolus3/drift#2324] [Q&A] [answered]
It is described [here](https://drift.simonbinder.eu/api/drift/migrationstrategy/oncreate) as 

> Executes when the database is opened for the first time.

And for me it seems like this method is run every time I start my application (and I connect to the DB on start). The naming sounds like it's only supposed to run when the DB is first created (i.e. it does not yet exist), but the description is slightly ambiguous.

Is my behavior buggy? It didn't use to run on every app start back with moor, at least not for me, only if the DB file didn't exist already.

**Comment by simolus3** [comment, marked as answer]
Uh oh, `onCreate` should definitely continue to only run when the database file is first created. `beforeOpen` is the callback invoked every time.

> Is my behavior buggy?

This sounds like a problem, yes. Is there any exception thrown by `onCreate`, or is the `schemaVersion` of your database set to `0`? Drift will invoke `onCreate` if the user version pragma in the database returns `0` and will set it to your current `schemaVersion` afterwards. If nothing fails there, `onCreate` should not run multiple times.

Could you run a custom select with `pragma user_version;` on the database and see what comes back?

**Reply by fgblomqvist** [reply]
It comes back as 0. However, I do now realize the exception that is showing is from my custom onCreate logic. I thought it was from a select. My bad!

If I remove my custom logic the `pragma user_version;` still comes back as zero but now when I restart the app it does no longer re-execute `onCreate`. Curious as to why the user_version does not get bumped, but at least if I fix the exception things should improve. Thanks!

**Reply by simolus3** [reply]
Huh, drift should actually run `PRAGMA user_version;` as well to determine whether `onCreate` runs or not. If you supply a `beforeOpen` callback to the `MigrationStrategy`, what does `details.versionBefore` evaluate to in there? Is it null/`0` as well?

**Reply by fgblomqvist** [reply]
`details.versionBefore` does work as expected. Tried reconnecting to the db in my IDE (IntelliJ) and now `pragma user_version;` returns the expected values. Not sure why that was needed, but perhaps it's related to some caching of some sort. I don't usually work with SQLite.

Either way, thanks for the support :)
